### PR TITLE
Support offline playback for downloaded HLS encrypted streams

### PR DESCRIPTION
**Pull request by ffelini**
With the current version of DTG lib [2.2.0](https://github.com/kaltura/playkit-dtg-android/releases/tag/v2.2.0) there is no way to play downloaded encrypted HLS streams because they contain the remote URI for the encryption key.  This URI is useless and inaccessible in offline mode.
It should be stored in the same folder along with the `.ts` and `.m3u8` files. The `variant.m3u8` file is updated accordingly using the local key path.

This PR introduces `AES-128` encryption key downloading for HLS streams so in offline playback the local key file path will be used.

The result of the key download shown in the console screenshots below

**Before**

screen shot 2017-08-07 at 4 50 34 pm

**After**

screen shot 2017-08-07 at 2 44 07 pm

### CODE REVIEW

**Comment by noamtamim** [comment]
Thanks, we will consider this contribution.

File: dtglib/src/main/java/com/kaltura/dtg/clear/HLSParser.java [hunk 1]
```diff
@@ -257,6 +263,42 @@ public String getPlaybackPath() {
...
+    }
+
+    /**
+     * @param segment
+     * @return DownloadTask for the remote encryption file. File name matches with the one stored
+     * in replaceRemoteEncryptionKeyWithLocal() method and the variant.m3u8 file
+     * @throws MalformedURLException
+     */
+    private DownloadTask createEncryptionKeyDownloadTask(Segment segment) throws MalformedURLException  {
+        String encryptionKeyFileName = createEncryptionKeyFileName(segment.encryptionKeyUri);
```
Comment (noamtamim) [inline dtglib/src/main/java/com/kaltura/dtg/clear/HLSParser.java]: Like master.m3u8 and variant.m3u8, there's only one encryption key -- so no real need to hash the URL. You can just call it "encryption.key".

File: dtglib/src/main/java/com/kaltura/dtg/clear/HLSParser.java [hunk 2]
```diff
@@ -244,6 +246,10 @@ public HlsMediaPlaylist getMediaPlaylist() {
                     segmentURL, segmentFile, segment.url, Utils.getHashedFileName(segment.url)));
 
             downloadTasks.add(new DownloadTask(segmentURL, segmentFile));
+
+            if (segment.isEncrypted) {
+                downloadTasks.add(createEncryptionKeyDownloadTask(segment));
```
Comment (noamtamim) [inline dtglib/src/main/java/com/kaltura/dtg/clear/HLSParser.java]: DownloadTasks are for the chunks. They are executed in a thread pool and have some overhead. I'd rather download the key directly, like how it's done for the m3u8 files.

File: dtglib/src/main/java/com/kaltura/dtg/clear/HLSParser.java (hunk 1 above)
Comment (ffelini) [inline dtglib/src/main/java/com/kaltura/dtg/clear/HLSParser.java]: I've had a think about that, but there may be cases where every chunk has a separate encryption key, which is very rare, but still.
Just found convenient to use that method which will generate unique file names for every encryption key remote url.

Comment (noamtamim) [inline dtglib/src/main/java/com/kaltura/dtg/clear/HLSParser.java]: Until now it didn't even support one... 

Do you have an example? How does such a playlist look like?

Comment (ffelini) [inline dtglib/src/main/java/com/kaltura/dtg/clear/HLSParser.java]: screen shot 2017-08-02 at 4 48 00 pm

Inspired from [here](https://support.theoplayer.com/hc/en-us/articles/206766939-AES-128-Content-Protection) 

I'm not sure if it's a feature supported only by the player from above and not a part of the HLS specification, but for me, it seems like it's a part of the HLS.
Also, I'm not actually sure if Exoplayer is supporting separate encryption keys for every chunk, and if kaltura play-kit has to send them to Exoplayer. 

Could you please check and let me know? Maybe this is really redundant for now

File: dtglib/src/main/java/com/kaltura/dtg/clear/HLSParser.java (hunk 2 above)
Comment (ffelini) [inline dtglib/src/main/java/com/kaltura/dtg/clear/HLSParser.java]: Fixed [here](https://github.com/kaltura/playkit-dtg-android/pull/15/commits/7677b91be8cf1e1ae741cfff20687493407b0f8b)

**Comment by noamtamim** [comment]
I used the sample you provided (from THEOplayer), wrapped in a custom master playlist (they only provide a single stream):
`https://noamtamim.com/random/hls/test-enc-aes/multi.m3u8`

Your code fails, because their key URLs are not absolute -- you need to add the base URL.

**Comment by ffelini** [comment]
I see, it fails... But I don't think DTG has to support non-absolute key URLs because this looks pretty tricky and is not specified anywhere. Even the current version of DTG + playkit will fail with such a `variant.m3u8` format because it should provide full URLs for the remote encryption keys. The issue is in this sample `https://noamtamim.com/random/hls/test-enc-aes/multi.m3u8`.
However, current branch implementation parses every line of "variant.m3u8" file and stores all the keys that will be found in there.

**Comment by ffelini** [comment]
@noamtamim Please confirm that encryption keys (if more than one is detected) are still needed to be stored using an obvious naming like: `encryption1.key`, `encryption2.key`, `encryption3.key`

**Comment by noamtamim** [comment]
@ffelini To support per-chunk keys as you mentioned, let's keep your original logic -- using hash of the URL.

**Comment by ffelini** [comment]
@noamtamim Great. Then that's all from my side. This PR looks to be ready. Thank you

**Comment by noamtamim** [comment]
@ffelini DTG **does** support non-absolute TS files; why is it harder to support non-absolute key files?
Anyway, I'll look into it. If it is indeed non trivial, I will list it as a known limitation.

**Comment by ffelini** [comment]
Ouch, sorry :) Now I got what I'm missing
This [commit](https://github.com/kaltura/playkit-dtg-android/pull/15/commits/0d0e31d6f7ec0931a2f78d3b71d5a4fbfffca851) ad the support for non-absolute keys.

**Comment by ffelini** [comment]
Unfortunately I'm on a holiday for two weeks. This PR will be quite for some time. If you have time to invest in this change I don't mind.

**Comment by ffelini** [comment]
@noamtamim  Requested changes have been implemented

**Comment by ffelini** [comment]
@noamtamim  Any chance to get this merged soon?

**Pull request merged**